## Transaction  Data Exploration 

In [ ]:
import pandas as pd
import numpy as np

In [894]:
quantium = pd.read_excel("/Users/thiernodicko/Desktop/Panda_files/QVI_transaction_data.xlsx", "sheet1")

In [895]:
# Printing the first 10 rows of the dataset
quantium

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0       43390          1            1000       1         5   
1       43599          1            1307     348        66   
2       43605          1            1343     383        61   
3       43329          2            2373     974        69   
4       43330          2            2426    1038       108   
...       ...        ...             ...     ...       ...   
264831  43533        272          272319  270088        89   
264832  43325        272          272358  270154        74   
264833  43410        272          272379  270187        51   
264834  43461        272          272379  270188        42   
264835  43365        272          272380  270189        74   

                                       PROD_NAME  PROD_QTY  TOT_SALES  
0         Natural Chip        Compny SeaSalt175g         2        6.0  
1                       CCs Nacho Cheese    175g         3        6.3  
2         Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3         Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4       Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8  
...                                          ...       ...        ...  
264831   Kettle Sweet Chilli And Sour Cream 175g         2       10.8  
264832             Tostitos Splash Of  Lime 175g         1        4.4  
264833                  Doritos Mexicana    170g         2        8.8  
264834   Doritos Corn Chip Mexican Jalapeno 150g         2        7.8  
264835             Tostitos Splash Of  Lime 175g         2        8.8  

[264836 rows x 8 columns]

In [896]:
# let view a summary of the dataset
quantium.describe()

DATE     STORE_NBR  LYLTY_CARD_NBR        TXN_ID  \
count  264836.000000  264836.00000    2.648360e+05  2.648360e+05   
mean    43464.036260     135.08011    1.355495e+05  1.351583e+05   
std       105.389282      76.78418    8.057998e+04  7.813303e+04   
min     43282.000000       1.00000    1.000000e+03  1.000000e+00   
25%     43373.000000      70.00000    7.002100e+04  6.760150e+04   
50%     43464.000000     130.00000    1.303575e+05  1.351375e+05   
75%     43555.000000     203.00000    2.030942e+05  2.027012e+05   
max     43646.000000     272.00000    2.373711e+06  2.415841e+06   

            PROD_NBR       PROD_QTY      TOT_SALES  
count  264836.000000  264836.000000  264836.000000  
mean       56.583157       1.907309       7.304200  
std        32.826638       0.643654       3.083226  
min         1.000000       1.000000       1.500000  
25%        28.000000       2.000000       5.400000  
50%        56.000000       2.000000       7.400000  
75%        85.000000       2.000000       9.200000  
max       114.000000     200.000000     650.000000

## Anomalies: 

First anomaly : Date column is in integer format, let change it to a date format

In [897]:
quantium['DATE'] = pd.to_datetime(quantium['DATE'], unit='D', origin='1899-12-30')

In [898]:
# Let show the format if it match to what we are looking for.
print(quantium['DATE'].dtype)

datetime64[ns]


In [899]:
# Printing the first 5 rows to see the update made.
quantium.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES  
0    Natural Chip        Compny SeaSalt175g         2        6.0  
1                  CCs Nacho Cheese    175g         3        6.3  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8

### Let examine the PROD_NAME column 

1- Since we are only looking at potato chips, let extract all digits from the prod_name column and create a new column to store them.

In [900]:
# Extract all digits and storing them to a variable called PACK_SIZE
PACK_SIZE = quantium['PROD_NAME'].str.extractall(r'(\d+)')

In [901]:
# Viewing the contain of the variable PACK_SIZE
PACK_SIZE.head(5)

0
  match     
0 0      175
1 0      175
2 0      170
3 0      175
4 0      150

We noticed that this time serie contain multiple index level, let drop one level before adding this column to our dataset

In [902]:
# Dropping the first level
quantium['PACK_SIZE'] = PACK_SIZE.droplevel(1)

In [903]:
# first 5 rows of the dataset
quantium.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   

                                  PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE  
0    Natural Chip        Compny SeaSalt175g         2        6.0       175  
1                  CCs Nacho Cheese    175g         3        6.3       175  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9       170  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0       175  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8       150

Extracting the unit from the PROD_NAME column. 

In [904]:
units = quantium['PROD_NAME'].str.extractall(r'(\d+)(\w+)')

In [905]:
# dropping the first level of the units dataframe
units = units.droplevel([1])

In [906]:
# Extracting the first column from the data
units = units.drop(0, axis=1)

In [907]:
# Showing the first 5 row of the data
units.head(5)

1
0  g
1  g
2  g
3  g
4  g

Let make sure that all units have the same symbol, if not, we can use the mode fuction to get the most frequent units.

In [908]:
units.mode()

1
0  g

In [909]:
units.at[0,1]

'g'

Append this unit to our PACK_SIZE column of the dataset

In [910]:
quantium = quantium.rename(columns={'PACK_SIZE': 'PACK_SIZE' + '_' + units.at[0,1]})

In [911]:
# Checking if the update was successful
quantium.head(10)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   
5 2019-05-19          4            4074    2982        57   
6 2019-05-16          4            4149    3333        16   
7 2019-05-16          4            4196    3539        24   
8 2018-08-20          5            5026    4525        42   
9 2018-08-18          7            7150    6900        52   

                                  PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
0    Natural Chip        Compny SeaSalt175g         2        6.0         175  
1                  CCs Nacho Cheese    175g         3        6.3         175  
2    Smiths Crinkle Cut  Chips Chicken 170g         2        2.9         170  
3    Smiths Chip Thinly  S/Cream&Onion 175g         5       15.0         175  
4  Kettle Tortilla ChpsHny&Jlpno Chili 150g         3       13.8         150  
5  Old El Paso Salsa   Dip Tomato Mild 300g         1        5.1         300  
6  Smiths Crinkle Chips Salt & Vinegar 330g         1        5.7         330  
7     Grain Waves         Sweet Chilli 210g         1        3.6         210  
8   Doritos Corn Chip Mexican Jalapeno 150g         1        3.9         150  
9     Grain Waves Sour    Cream&Chives 210G         2        7.2         210

Now let remove all extracted digits and units from the PROD_NAME column and replace them by an empty space.

In [912]:
quantium['PROD_NAME'] = quantium['PROD_NAME'].replace([r'(\d+)(\w+)'], '', regex=True)

In [913]:
# Verifiying the update made
quantium.head(5)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0 2018-10-17          1            1000       1         5   
1 2019-05-14          1            1307     348        66   
2 2019-05-20          1            1343     383        61   
3 2018-08-17          2            2373     974        69   
4 2018-08-18          2            2426    1038       108   

                              PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
0    Natural Chip        Compny SeaSalt         2        6.0         175  
1                  CCs Nacho Cheese             3        6.3         175  
2    Smiths Crinkle Cut  Chips Chicken          2        2.9         170  
3    Smiths Chip Thinly  S/Cream&Onion          5       15.0         175  
4  Kettle Tortilla ChpsHny&Jlpno Chili          3       13.8         150

Removing all special characters from the prod_name column

In [914]:
quantium['PROD_NAME'] = quantium['PROD_NAME'].replace(r'[^\w\s]', '', regex=True)

Getting word from Prod_name column, splitting them into words, finding the frequency of each word and ordering them by their frequency from highest to lowest frequency. 

In [915]:
# collecting word from the prod_name column
product_words = merged_data['PROD_NAME']

In [916]:
# Creating a time serie
product_words = pd.Series(' '.join(product_words).split())

In [917]:
# finding the frequency of each unique word
product_freq = pd.Series(product_words).value_counts().reset_index()

In [918]:
# renaming columns
product_freq.columns = ['words','frequency']

In [919]:
product_freq

words  frequency
0       Chips      49770
1      Kettle      41288
2        Salt      27976
3      Cheese      27890
4      Smiths      27390
..        ...        ...
175  FrchOnin       1432
176        Pc       1431
177    Garden       1419
178       NCC       1419
179     Fries       1418

[180 rows x 2 columns]

Since we are only interest in the chips category, let remove any salsa products from the dataset

In [920]:
# creating a column Salsa that will return true if salsa appears in the product category, false otherwise
quantium['SALSA'] = quantium['PROD_NAME'].str.lower().str.contains('salsa')

In [921]:
# Printing rows where the product salsa appears
quantium[quantium.SALSA == True]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
5      2019-05-19          4            4074    2982        57   
25     2019-05-15         39           39144   35506        57   
32     2019-05-20         45           45127   41122        64   
44     2018-08-18         56           56013   50090        39   
63     2019-05-15         82           82480   82047       101   
...           ...        ...             ...     ...       ...   
264675 2019-04-20        265          265103  263419        59   
264678 2019-03-30        265          265111  263428        35   
264719 2018-10-28        266          266278  264104        39   
264734 2019-01-11        267          267324  264374        41   
264780 2019-01-10        269          269222  266382        64   

                                   PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  \
5       Old El Paso Salsa   Dip Tomato Mild          1        5.1         300   
25      Old El Paso Salsa   Dip Tomato Mild          1        5.1         300   
32       Red Rock Deli SR    Salsa  Mzzrlla          2        5.4         150   
44         Smiths Crinkle Cut  Tomato Salsa          1        2.6         150   
63               Doritos Salsa       Medium          1        2.6         300   
...                                      ...       ...        ...         ...   
264675   Old El Paso Salsa   Dip Tomato Med          1        5.1         300   
264678            Woolworths Mild     Salsa          1        1.5         300   
264719     Smiths Crinkle Cut  Tomato Salsa          1        2.6         150   
264734                  Doritos Salsa Mild           1        2.6         300   
264780   Red Rock Deli SR    Salsa  Mzzrlla          2        5.4         150   

        SALSA  
5        True  
25       True  
32       True  
44       True  
63       True  
...       ...  
264675   True  
264678   True  
264719   True  
264734   True  
264780   True  

[18094 rows x 10 columns]

Remove all rows where the word "salsa" appears, then drop the column SALSA

In [922]:
quantium = quantium[~quantium['SALSA']]

In [923]:
# dropping the column SALSA
quantium = quantium.drop(columns='SALSA')

In [924]:
# Checking update made
quantium.head(10)

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
0  2018-10-17          1            1000       1         5   
1  2019-05-14          1            1307     348        66   
2  2019-05-20          1            1343     383        61   
3  2018-08-17          2            2373     974        69   
4  2018-08-18          2            2426    1038       108   
6  2019-05-16          4            4149    3333        16   
7  2019-05-16          4            4196    3539        24   
8  2018-08-20          5            5026    4525        42   
9  2018-08-18          7            7150    6900        52   
10 2019-05-17          7            7215    7176        16   

                              PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
0    Natural Chip        Compny SeaSalt         2        6.0         175  
1                  CCs Nacho Cheese             3        6.3         175  
2    Smiths Crinkle Cut  Chips Chicken          2        2.9         170  
3      Smiths Chip Thinly  SCreamOnion          5       15.0         175  
4   Kettle Tortilla ChpsHnyJlpno Chili          3       13.8         150  
6   Smiths Crinkle Chips Salt  Vinegar          1        5.7         330  
7     Grain Waves         Sweet Chilli          1        3.6         210  
8   Doritos Corn Chip Mexican Jalapeno          1        3.9         150  
9      Grain Waves Sour    CreamChives          2        7.2         210  
10  Smiths Crinkle Chips Salt  Vinegar          1        5.7         330

In [925]:
# Checking for null values
quantium.isnull().values.any()

np.False_

In [926]:
# Setting date column as the index column
quantium.set_index('DATE', inplace=True)


In [927]:
# Summary of the dataset
quantium.describe()

STORE_NBR  LYLTY_CARD_NBR        TXN_ID       PROD_NBR  \
count  246742.000000    2.467420e+05  2.467420e+05  246742.000000   
mean      135.051098    1.355310e+05  1.351311e+05      56.351789   
std        76.787096    8.071528e+04  7.814772e+04      33.695428   
min         1.000000    1.000000e+03  1.000000e+00       1.000000   
25%        70.000000    7.001500e+04  6.756925e+04      26.000000   
50%       130.000000    1.303670e+05  1.351830e+05      53.000000   
75%       203.000000    2.030840e+05  2.026538e+05      87.000000   
max       272.000000    2.373711e+06  2.415841e+06     114.000000   

            PROD_QTY      TOT_SALES  
count  246742.000000  246742.000000  
mean        1.908062       7.321322  
std         0.659831       3.077828  
min         1.000000       1.700000  
25%         2.000000       5.800000  
50%         2.000000       7.400000  
75%         2.000000       8.800000  
max       200.000000     650.000000

We noticed that there are not null values, however the product quatity column appears to have an outlier which we should investigate further.

In [928]:
# Filter the product quantity column to find the outlier
quantium[quantium.PROD_QTY == 200]

STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
DATE                                                      
2018-08-19        226          226000  226201         4   
2019-05-20        226          226000  226210         4   

                               PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
DATE                                                                       
2018-08-19  Dorito Corn Chp     Supreme        200      650.0         380  
2019-05-20  Dorito Corn Chp     Supreme        200      650.0         380

Let check if the above customer has made other transactions

In [929]:
quantium[quantium.LYLTY_CARD_NBR == 226000]

STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
DATE                                                      
2018-08-19        226          226000  226201         4   
2019-05-20        226          226000  226210         4   

                               PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
DATE                                                                       
2018-08-19  Dorito Corn Chp     Supreme        200      650.0         380  
2019-05-20  Dorito Corn Chp     Supreme        200      650.0         380

Its looks like this customer has only had the two transactions over the year and is not an ordinary retail customer.The customer might be bying chips for commercial purposes instead. Let remove this loyalty card number from the dataset.

In [930]:
# Removing loyalty card number 226000 from the dataset
quantium = quantium[quantium['LYLTY_CARD_NBR'] != 226000]

In [931]:
# Checking if the update was successful
quantium

STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
DATE                                                      
2018-10-17          1            1000       1         5   
2019-05-14          1            1307     348        66   
2019-05-20          1            1343     383        61   
2018-08-17          2            2373     974        69   
2018-08-18          2            2426    1038       108   
...               ...             ...     ...       ...   
2019-03-09        272          272319  270088        89   
2018-08-13        272          272358  270154        74   
2018-11-06        272          272379  270187        51   
2018-12-27        272          272379  270188        42   
2018-09-22        272          272380  270189        74   

                                      PROD_NAME  PROD_QTY  TOT_SALES  \
DATE                                                                   
2018-10-17   Natural Chip        Compny SeaSalt         2        6.0   
2019-05-14                 CCs Nacho Cheese             3        6.3   
2019-05-20   Smiths Crinkle Cut  Chips Chicken          2        2.9   
2018-08-17     Smiths Chip Thinly  SCreamOnion          5       15.0   
2018-08-18  Kettle Tortilla ChpsHnyJlpno Chili          3       13.8   
...                                         ...       ...        ...   
2019-03-09  Kettle Sweet Chilli And Sour Cream          2       10.8   
2018-08-13            Tostitos Splash Of  Lime          1        4.4   
2018-11-06                 Doritos Mexicana             2        8.8   
2018-12-27  Doritos Corn Chip Mexican Jalapeno          2        7.8   
2018-09-22            Tostitos Splash Of  Lime          2        8.8   

           PACK_SIZE_g  
DATE                    
2018-10-17         175  
2019-05-14         175  
2019-05-20         170  
2018-08-17         175  
2018-08-18         150  
...                ...  
2019-03-09         175  
2018-08-13         175  
2018-11-06         170  
2018-12-27         150  
2018-09-22         175  

[246740 rows x 8 columns]

In [932]:
# Summary of the data
quantium.describe()

STORE_NBR  LYLTY_CARD_NBR        TXN_ID       PROD_NBR  \
count  246740.000000    2.467400e+05  2.467400e+05  246740.000000   
mean      135.050361    1.355303e+05  1.351304e+05      56.352213   
std        76.786971    8.071520e+04  7.814760e+04      33.695235   
min         1.000000    1.000000e+03  1.000000e+00       1.000000   
25%        70.000000    7.001500e+04  6.756875e+04      26.000000   
50%       130.000000    1.303670e+05  1.351815e+05      53.000000   
75%       203.000000    2.030832e+05  2.026522e+05      87.000000   
max       272.000000    2.373711e+06  2.415841e+06     114.000000   

            PROD_QTY      TOT_SALES  
count  246740.000000  246740.000000  
mean        1.906456       7.316113  
std         0.342499       2.474897  
min         1.000000       1.700000  
25%         2.000000       5.800000  
50%         2.000000       7.400000  
75%         2.000000       8.800000  
max         5.000000      29.500000

In [933]:
# Counting the number of transaction by date
count_by_date = merged_data.groupby('DATE').size()

In [934]:
count_by_date

DATE
2018-07-01    663
2018-07-02    650
2018-07-03    674
2018-07-04    669
2018-07-05    660
             ... 
2019-06-26    657
2019-06-27    669
2019-06-28    673
2019-06-29    703
2019-06-30    704
Length: 364, dtype: int64

There's only 364 rows, meaning only 364 dates which indicates a missing date. Let's
create a sequence of dates from 1 Jul 2018 to 30 Jun 2019 and use this to create a
chart of number of transactions over time to find the missing date.

In [935]:
# Sequence of date from 1 Jul 2018 to 30 Jun 2019
transaction_date = pd.date_range(start='2018-07-01', end='2019-06-30', freq='D')

In [936]:
# Printing the dates
transaction_date

DatetimeIndex(['2018-07-01', '2018-07-02', '2018-07-03', '2018-07-04',
               '2018-07-05', '2018-07-06', '2018-07-07', '2018-07-08',
               '2018-07-09', '2018-07-10',
               ...
               '2019-06-21', '2019-06-22', '2019-06-23', '2019-06-24',
               '2019-06-25', '2019-06-26', '2019-06-27', '2019-06-28',
               '2019-06-29', '2019-06-30'],
              dtype='datetime64[ns]', length=365, freq='D')

Let reset the date index, sort the date, then show missing date

In [937]:
# Resetting the date index.
quantium = quantium.reset_index()

In [938]:
# Sorting the data by Date
quantium.sort_values(by='DATE')

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR  \
8572   2018-07-01         88           88140   86914        25   
144861 2018-07-01         60           60276   57330         3   
168994 2018-07-01        199          199014  197623       104   
214152 2018-07-01         35           35052   31630        11   
97432  2018-07-01         72           72104   71038        20   
...           ...        ...             ...     ...       ...   
9601   2019-06-30        112          112141  114611        98   
105465 2019-06-30        207          207155  205513        99   
213436 2019-06-30         10           10140    9882        12   
213283 2019-06-30          6            6258    6047        29   
244804 2019-06-30        183          183196  185975        22   

                                   PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
8572              Pringles SourCream  Onion          2        7.4         134  
144861   Kettle Sensations   Camembert  Fig          2        9.2         150  
168994  Infuzions Thai SweetChili PotatoMix          2        7.6         110  
214152                  RRD Pc Sea Salt              1        3.0         165  
97432           Doritos Cheese      Supreme          2       11.4         330  
...                                      ...       ...        ...         ...  
9601       NCC Sour Cream     Garden Chives          2        6.0         175  
105465          Pringles Sthrn FriedChicken          2        7.4         134  
213436    Natural Chip Co     Tmato HrbSpce          2        6.0         175  
213283            French Fries Potato Chips          1        3.0         175  
244804    Thins Chips         Originl saltd          2        6.6         175  

[246740 rows x 9 columns]

In [939]:
# showing missing date
transaction_date[~transaction_date.isin(quantium['DATE'])]

DatetimeIndex(['2018-12-25'], dtype='datetime64[ns]', freq='D')

We noticed that there is one missing date. Now let create a column of dates that includes every day from 1 Jul 2018 to 30 Jun 2019, and join it onto the data to fill in the missing day.

In [940]:
# Creating a dataframe for the new sequence date created
complete_date = pd.DataFrame(transaction_date, columns=['DATE'])

In [941]:
# Joining the sequence date with the existing date
quantium = complete_date.merge(quantium, on='DATE', how='left')

In [942]:
# Verify if there still a missing date in the data
transaction_date[~transaction_date.isin(quantium['DATE'])]

DatetimeIndex([], dtype='datetime64[ns]', freq='D')

In [943]:
# Let verify if there are some null values
quantium.isnull().values.any()

np.True_

In [944]:
quantium[quantium.PROD_NAME.isnull()]

DATE  STORE_NBR  LYLTY_CARD_NBR  TXN_ID  PROD_NBR PROD_NAME  \
120465 2018-12-25        NaN             NaN     NaN       NaN       NaN   

        PROD_QTY  TOT_SALES PACK_SIZE_g  
120465       NaN        NaN         NaN

We have a entire row with null value, let get ride of it. 

In [945]:
quantium.dropna(inplace=True)

In [946]:
# Rechecking for nan values.
quantium[quantium.PROD_NAME.isnull()]

Empty DataFrame
Columns: [DATE, STORE_NBR, LYLTY_CARD_NBR, TXN_ID, PROD_NBR, PROD_NAME, PROD_QTY, TOT_SALES, PACK_SIZE_g]
Index: []

No missing date, no null values from our data. Now let move on to verify the output from the pack size we created earlier, and to creating other future such as brand of chips which can be done by extracting the first word of each row of the product name column.

In [947]:
quantium.sort_values(by="PACK_SIZE_g")

DATE  STORE_NBR  LYLTY_CARD_NBR    TXN_ID  PROD_NBR  \
77628  2018-10-23      256.0        256448.0  255636.0       2.0   
192679 2019-04-12      236.0        236142.0  239408.0     104.0   
91995  2018-11-14       72.0         72165.0   71319.0      31.0   
92008  2018-11-14      238.0        238125.0  242382.0     104.0   
92013  2018-11-14      271.0        271162.0  269165.0      87.0   
...           ...        ...             ...       ...       ...   
9904   2018-07-15      180.0        180092.0  181569.0      95.0   
4148   2018-07-07      157.0        157057.0  158757.0      95.0   
30815  2018-08-15      229.0        229236.0  231796.0      95.0   
138814 2019-01-22      178.0        178040.0  177814.0      55.0   
93762  2018-11-16      100.0        100003.0   98331.0      95.0   

                                   PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  
77628      Cobs Popd Sour Crm  Chives Chips        2.0        7.6         110  
192679  Infuzions Thai SweetChili PotatoMix        2.0        7.6         110  
91995     Infzns Crn Crnchers Tangy Gcamole        2.0        7.6         110  
92008   Infuzions Thai SweetChili PotatoMix        2.0        7.6         110  
92013    Infuzions BBQ Rib   Prawn Crackers        2.0        7.6         110  
...                                      ...       ...        ...         ...  
9904    Sunbites Whlegrn    Crisps FrchOnin        2.0        3.4          90  
4148    Sunbites Whlegrn    Crisps FrchOnin        2.0        3.4          90  
30815   Sunbites Whlegrn    Crisps FrchOnin        1.0        1.7          90  
138814      Snbts Whlgrn Crisps CheddrMstrd        2.0        3.4          90  
93762   Sunbites Whlegrn    Crisps FrchOnin        2.0        3.4          90  

[246740 rows x 9 columns]

In [948]:
# Creating a column for the brand name.
quantium.insert(9, "BRAND_NAME",quantium['PROD_NAME'].str.split().str.get(0), True)

In [949]:
quantium

DATE  STORE_NBR  LYLTY_CARD_NBR    TXN_ID  PROD_NBR  \
0      2018-07-01       47.0         47142.0   42540.0      14.0   
1      2018-07-01       55.0         55073.0   48884.0      99.0   
2      2018-07-01       55.0         55073.0   48884.0      91.0   
3      2018-07-01       58.0         58351.0   54374.0     102.0   
4      2018-07-01       68.0         68193.0   65598.0      44.0   
...           ...        ...             ...       ...       ...   
246736 2019-06-30      242.0        242159.0  246222.0      36.0   
246737 2019-06-30      244.0        244213.0  247339.0      93.0   
246738 2019-06-30      256.0        256018.0  255130.0     105.0   
246739 2019-06-30      257.0        257079.0  256218.0      71.0   
246740 2019-06-30      265.0        265006.0  263307.0     106.0   

                                  PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  \
0        Smiths Crnkle Chip  Orgnl Big Bag        2.0       11.8         380   
1              Pringles Sthrn FriedChicken        2.0        7.4         134   
2                      CCs Tasty Cheese           2.0        4.2         175   
3         Kettle Mozzarella   Basil  Pesto        2.0       10.8         175   
4                 Thins Chips Light  Tangy        2.0        6.6         175   
...                                     ...       ...        ...         ...   
246736                       Kettle Chilli        2.0       10.8         175   
246737  Doritos Corn Chip Southern Chicken        2.0        7.8         150   
246738           Woolworths Cheese   Rings        2.0        3.6         190   
246739          Twisties Cheese     Burger        2.0        8.6         250   
246740   Natural ChipCo      Hony Soy Chckn       1.0        3.0         175   

        BRAND_NAME  
0           Smiths  
1         Pringles  
2              CCs  
3           Kettle  
4            Thins  
...            ...  
246736      Kettle  
246737     Doritos  
246738  Woolworths  
246739    Twisties  
246740     Natural  

[246740 rows x 10 columns]

Let show all the unique entries of the brand name column

In [954]:
quantium.BRAND_NAME.unique()

array(['Smiths', 'Pringles', 'CCs', 'Kettle', 'Thins', 'Infuzions',
       'GrnWves', 'RRD', 'Twisties', 'Red', 'Cheezels', 'Tostitos', 'WW',
       'Grain', 'Doritos', 'Cobs', 'Natural', 'Burger', 'Dorito',
       'Cheetos', 'Smith', 'French', 'NCC', 'Infzns', 'Sunbites',
       'Tyrrells', 'Snbts', 'Woolworths'], dtype=object)

Some of the brand names look like they are of the same brands, such as RED and RRD which are both Red Rock Deli Chips. Let's combine these together and make any additional brand adjustments.

In [955]:
# Creating a function that will serve to join all brands with the same type
def replace_brandname(line):
    name = line['BRAND_NAME']
    if name == "Infzns":
        return "Infuzions"
    elif name == "Red":
        return "Red Rock Deli"
    elif name == "RRD":
        return "Red Rock Deli"
    elif name == "Grain":
        return "Grain Waves"
    elif name == "GrnWves":
        return "Grain Waves"
    elif name == "Snbts":
        return "Sunbites"
    elif name == "Natural":
        return "Natural Chip Co"
    elif name == "NCC":
        return "Natural Chip Co"
    elif name == "WW":
        return "Woolworths"
    elif name == "Smith":
        return "Smiths"
    elif name == "Dorito":
        return "Doritos"
    else:
        return name 

In [958]:
# Then apply the function to clean the brand names
quantium["BRAND_NAME"] = quantium.apply(lambda line: replace_brandname(line), axis=1)

In [959]:
# Verify the change made.
quantium.BRAND_NAME.unique()

array(['Smiths', 'Pringles', 'CCs', 'Kettle', 'Thins', 'Infuzions',
       'Grain Waves', 'Red Rock Deli', 'Twisties', 'Cheezels', 'Tostitos',
       'Woolworths', 'Doritos', 'Cobs', 'Natural Chip Co', 'Burger',
       'Cheetos', 'French', 'Sunbites', 'Tyrrells'], dtype=object)

The above brand name seemed reasonable without duplicate.

### Customer Data Exploration

Importing customer data 

In [960]:
customer = pd.read_csv("/Users/thiernodicko/Desktop/Panda_files/QVI_purchase_behaviour.csv")

In [961]:
# looking a the first 10 rows of the customer data
customer.head(10)

LYLTY_CARD_NBR               LIFESTAGE PREMIUM_CUSTOMER
0            1000   YOUNG SINGLES/COUPLES          Premium
1            1002   YOUNG SINGLES/COUPLES       Mainstream
2            1003          YOUNG FAMILIES           Budget
3            1004   OLDER SINGLES/COUPLES       Mainstream
4            1005  MIDAGE SINGLES/COUPLES       Mainstream
5            1007   YOUNG SINGLES/COUPLES           Budget
6            1009            NEW FAMILIES          Premium
7            1010   YOUNG SINGLES/COUPLES       Mainstream
8            1011   OLDER SINGLES/COUPLES       Mainstream
9            1012          OLDER FAMILIES       Mainstream

In [962]:
# Check a summary of the customer data
customer.describe()

LYLTY_CARD_NBR
count    7.263700e+04
mean     1.361859e+05
std      8.989293e+04
min      1.000000e+03
25%      6.620200e+04
50%      1.340400e+05
75%      2.033750e+05
max      2.373711e+06

In [963]:
# Check for null values in the dataset
customer.isnull().values.any()

np.False_

In [964]:
# Let rename the column 'PREMIUM_CUSTOMER' to 'MEMBER_TYPE' 

customer = customer.rename(columns={"PREMIUM_CUSTOMER" : "MEMBER_TYPE"})

In [965]:
customer

LYLTY_CARD_NBR               LIFESTAGE MEMBER_TYPE
0                1000   YOUNG SINGLES/COUPLES     Premium
1                1002   YOUNG SINGLES/COUPLES  Mainstream
2                1003          YOUNG FAMILIES      Budget
3                1004   OLDER SINGLES/COUPLES  Mainstream
4                1005  MIDAGE SINGLES/COUPLES  Mainstream
...               ...                     ...         ...
72632         2370651  MIDAGE SINGLES/COUPLES  Mainstream
72633         2370701          YOUNG FAMILIES  Mainstream
72634         2370751          YOUNG FAMILIES     Premium
72635         2370961          OLDER FAMILIES      Budget
72636         2373711   YOUNG SINGLES/COUPLES  Mainstream

[72637 rows x 3 columns]

In [966]:
# Let check the entries of both column Lifestage and member_type in the dataset
customer['LIFESTAGE'].unique()

array(['YOUNG SINGLES/COUPLES', 'YOUNG FAMILIES', 'OLDER SINGLES/COUPLES',
       'MIDAGE SINGLES/COUPLES', 'NEW FAMILIES', 'OLDER FAMILIES',
       'RETIREES'], dtype=object)

In [967]:
customer['MEMBER_TYPE'].unique()

array(['Premium', 'Mainstream', 'Budget'], dtype=object)

The data looks fine: no null or missing values and no duplicate.

Merging Transaction Data and Customer Data

In [970]:
# Join the customer and transaction datasets, and sort transactons by date 
full_data = quantium.set_index('LYLTY_CARD_NBR').join(customer.set_index('LYLTY_CARD_NBR'))
full_data = full_data.reset_index()
full_data = full_data.sort_values(by='DATE').reset_index(drop=True)


In [971]:
full_data

LYLTY_CARD_NBR       DATE  STORE_NBR    TXN_ID  PROD_NBR  \
0              47142.0 2018-07-01       47.0   42540.0      14.0   
1              28024.0 2018-07-01       28.0   24672.0     113.0   
2              33107.0 2018-07-01       33.0   30115.0     105.0   
3              40130.0 2018-07-01       40.0   36721.0      63.0   
4              45109.0 2018-07-01       45.0   41028.0      79.0   
...                ...        ...        ...       ...       ...   
246735        107109.0 2019-06-30      107.0  108968.0     113.0   
246736        119271.0 2019-06-30      119.0  123199.0       9.0   
246737        122230.0 2019-06-30      122.0  125994.0     106.0   
246738        213140.0 2019-06-30      213.0  212816.0      90.0   
246739        265006.0 2019-06-30      265.0  263307.0     106.0   

                                 PROD_NAME  PROD_QTY  TOT_SALES PACK_SIZE_g  \
0       Smiths Crnkle Chip  Orgnl Big Bag        2.0       11.8         380   
1                         Twisties Chicken       2.0        9.2         270   
2               Woolworths Cheese   Rings        2.0        3.6         190   
3                 Kettle  Swt Pot Sea Salt       2.0        8.4         135   
4        Smiths Chip Thinly  CutSaltVinegr       2.0        6.0         175   
...                                    ...       ...        ...         ...   
246735                    Twisties Chicken       2.0        9.2         270   
246736  Kettle Tortilla ChpsBtrootRicotta        2.0        9.2         150   
246737  Natural ChipCo      Hony Soy Chckn       2.0        6.0         175   
246738       Tostitos Smoked     Chipotle        2.0        8.8         175   
246739  Natural ChipCo      Hony Soy Chckn       1.0        3.0         175   

             BRAND_NAME               LIFESTAGE MEMBER_TYPE  
0                Smiths  MIDAGE SINGLES/COUPLES      Budget  
1              Twisties                RETIREES     Premium  
2            Woolworths                RETIREES     Premium  
3                Kettle                RETIREES     Premium  
4                Smiths                RETIREES     Premium  
...                 ...                     ...         ...  
246735         Twisties   OLDER SINGLES/COUPLES      Budget  
246736           Kettle   OLDER SINGLES/COUPLES      Budget  
246737  Natural Chip Co   OLDER SINGLES/COUPLES      Budget  
246738         Tostitos   OLDER SINGLES/COUPLES     Premium  
246739  Natural Chip Co   YOUNG SINGLES/COUPLES     Premium  

[246740 rows x 12 columns]

In [972]:
# Checkin for null values.
full_data.isnull().values.any()

np.False_

The final data looks reasonable, hence let export it to csv file for further analysis.

In [973]:
full_data.to_csv('QVI_finaldata.csv')

### Next: I will use Power BI for data visualization and SQL to answer some business questions. 